### ISSUE TITLE
Java app: High CPU after blue-green deployment

### ISSUE THREAD

**Issue by jtheuer** [boxfuse/boxfuse-issues#182]
We have a high traffic api using boxfuse. We noticed that right after a blue-green deployment the new instances have much higher CPU load than the old instances:

<img width="911" alt="cpu/memory" src="https://user-images.githubusercontent.com/96364/30383248-9443238c-98a1-11e7-9227-8660bd910b46.png">

You see in this graph the CPU measured by newrelic of a boxfuse/java app on a m4.large instance: two instances are running and two new instances are starting.
You see the CPU of the original instances drop at some point (taken out of ELB by boxfuse) but the new instances still struggle with the same load at 200% (2 cores @ 100%). The load is so high that auto-scaling even triggers a 3rd new instances (yellow).

We don't have any pre computations or caching in our user code and also GC doesn't show anything suspicious (<1%)  so we _think_ this is caused by java not being warm (i.e. not all code optimisations done, yet) but I never thought the effect would be so high? Do you have experiences with that? Any ideas what could be the reason? NewRelic shows only "java" and "nrsysmond.x64" running, I guess there is nothing boxfuse-specific running after start, correct?

Does anybody have had similar java experiences (with boxfuse or other deployment tools)?

Any idea how the blue-green deployment can be improved? Maybe start more instances than currently are running or add a transition time before old instances are deleted? (Running a mix of old and new instances might not work for all customers but would be ok for us)

Thanks,

Jan

**Comment by axelfontaine** [comment]
From our side I can say that we have not experienced this behavior so far.

I can confirm that on the instance there will only be the following processes running:
- an init shell script that launches all other processes described below
- an acpi daemon for responding to power events
- a ntp daemon to prevent clock drift
- your app (a JVM in this case)
- when using New Relic, the New Relic servers agent
- when using CloudWatch Logs, 2 instances of https://github.com/boxfuse/cloudwatchlogs-agent (one for stdout and one for stderr)

**Comment by jtheuer** [comment]
Thanks for the confirmation!
I don't think any of the processes your listed is responsible for the high CPU load but it is rather the JVM. So either it is NewRelic, something in our code we still couldn't find or a "cold start problem".

Does anybody here has experience with JVM starts? How much optimisation does the JVM on start and could make such a difference (50% CPU vs 100% CPU under high load).

What do you think about a configurable grace period (wait between new instances are online and old instances are removed)? Other ideas?

**Comment by axelfontaine** [comment]
Some areas you could explore to reduce the initial CPU usage peak are:

- lazy initialization of services/caches/...
- deferred start of scheduled jobs (wait X seconds before the initial run)
- disabling JVM bytecode verification
- also check if there are opportunities to move computation/instrumentalization from run time to build time

**Comment by jtheuer** [comment]
It took us quite a while to get to the bottom of this since it wasn't easily reproducible. In the end it was likely that the fresh instances couldn't handle the same load as the old instances. This was only a problem when instances were already close to their scaling limit.

One possible fix would have been to lower the scaling thresholds, doing rolling/batch updates or trying the new https://docs.aws.amazon.com/autoscaling/ec2/userguide/as-scaling-target-tracking.html but for us, luckily we were already successful with changing the instances types from m4.large to c4.xlarge (after evaluating all *4*large types with production load).

So bottom line if you run into this problem: Check scaling and instances types.
